pkg/reconciler/instances/istio/action.go

```diff
@@ -23,20 +40,174 @@ func (a *ReconcileAction) Run(version, profile string, config map[string]interfa
 		return err
 	}
 
-	performer, err := actions.NewDefaultIstioPerformer(context.KubeClient.Kubeconfig(), manifest.Manifest, context.KubeClient, context.Logger, &istioctl.DefaultCommander{})
+	ver, err := a.performer.Version(context.WorkspaceFactory, version, istioChart, context.KubeClient.Kubeconfig(), context.Logger)
+	if err != nil {
+		return errors.Wrap(err, "Could not fetch Istio version")
+	}
+
+	context.Logger.Infof("Detected versions: istioctl %s, pilot version: %s, data plane version: %s", ver.ClientVersion, ver.PilotVersion, ver.DataPlaneVersion)
+
+	if shouldInstall(ver) {
+		context.Logger.Info("No Istio version was detected on the cluster, performing installation...")
+
+		err = a.performer.Install(context.KubeClient.Kubeconfig(), manifest.Manifest, context.Logger)
+		if err != nil {
+			return errors.Wrap(err, "Could not install Istio")
+		}
+
+		err = a.performer.PatchMutatingWebhook(context.KubeClient, context.Logger)
+		if err != nil {
+			return errors.Wrap(err, "Could not patch MutatingWebhookConfiguration")
+		}
+	} else {
+		if !isClientVersionAcceptable(ver) {
```

Comment:
Maybe you think about moving that several checks into a new function li,e "isUpgradePossible" to make it more expressive